python/dgl/nn/pytorch/explain/gnnexplainer.py

```diff
@@ -425,3 +398,395 @@ def explain_graph(self, graph, feat, **kwargs):
+    lr : float, optional
+        The learning rate to use, default to 0.01.
+    num_epochs : int, optional
+        The number of epochs to train.
+    alpha1 : float, optional
+        A higher value will make the explanation edge masks more sparse by decreasing
+        the sum of the edge mask.
+    alpha2 : float, optional
+        A higher value will make the explanation edge masks more sparse by decreasing
+        the entropy of the edge mask.
+    beta1 : float, optional
+        A higher value will make the explanation node feature masks more sparse by
+        decreasing the mean of the node feature mask.
+    beta2 : float, optional
+        A higher value will make the explanation node feature masks more sparse by
+        decreasing the entropy of the node feature mask.
+    log : bool, optional
+        If True, it will log the computation process, default to True.
+    """
+    
+    def __init__(self, 
+                 model, 
+                 num_hops, 
+                 lr=0.01, 
+                 num_epochs=100, 
+                 *,
+                 alpha1=0.005,
+                 alpha2=1.0,
+                 beta1=1.0,
+                 beta2=0.1,
+                 log=True):
+        super(HeteroGNNExplainer, self).__init__()
+        self.model = model
+        self.num_hops = num_hops
+        self.lr = lr
+        self.num_epochs = num_epochs
+        self.alpha1 = alpha1
+        self.alpha2 = alpha2
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.log = log
+
+    def _init_masks(self, graph, feat):
+        r"""Initialize learnable feature and edge mask.
+        Parameters
+        ----------
+        graph : DGLHeteroGraph
+            Input graph.
+        feat : Dictionary of Tensor
+            Input node features.
+        Returns
+        -------
+        feat_masks : Dictionary of Tensor
+            Dictionary of { ntypes: features }
+            Feature mask of shape :math:`(1, D)`, where :math:`D`
+            is the feature size.
+        edge_masks : Dictionary of Tensor
+            Dictionary of { canonical_etypes: features }
+            Edge mask of shape :math:`(E)`, where :math:`E` is the
+            number of edges.
+        """
+        device = graph.device
+        feat_masks = {}
+        for node_type, feature in feat.items():
+            if len(feature.size()) == 1:
+                feat_masks[node_type] = nn.Parameter(torch.zeros(1, 1, device=device))
+            else:
+                std = 0.1
+                num_nodes, feat_size = feature.size()
+                feat_masks[node_type] = nn.Parameter(torch.randn(1, feat_size, device=device) * std)
+
+        edge_masks = {}
+        for canonical_etype in graph.canonical_etypes:
+            src = canonical_etype[0]
+            num_nodes = graph.number_of_nodes(src)
+            num_edges = graph.number_of_edges(canonical_etype)
+            std = nn.init.calculate_gain('relu')
+            if num_nodes > 0:
```

Comment:
Do we need to handle this case? If `num_edges` is positive, `num_nodes` will also be positive.

Reply:
From our dataset, we have seen that there are heterographs where a type of node is not present, for example, no socket nodes present (no internet). Some background information, our dataset schema has three node types: `process, file, and socket`, and four edge types: `PROC_CREATE, READ, WRITE, FILE_EXEC`. So, there will exist a canonical edge type: `'Process','READ','Socket': []`, and while initializing a mask for this relation, it will get a divide by 0 error (since no socket nodes are present).